cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1892,6 +1896,9 @@ class ObjectOutputStream final : public io::OutputStream {
     }
     // Notify completion
     if (--state->parts_in_progress == 0) {
+      // GH-41862: avoid potential deadlock if the Future's callback is called
+      // with the mutex taken.
+      lock.unlock();
       state->pending_parts_completed.MarkFinished(state->status);
```

Comment:
Now it's possible for this object to go from "Finished" to "not-finished". There might be logic relying on the state machine converging to the finished state and staying there.

I might do a full review of this class at some point.

Reply:
> Now it's possible for this object to go from "Finished" to "not-finished".

Where is it possible?

Reply:
```c++
auto fut = state->pending_parts_completed;
lock.unlock();
fut.MarkFinished(state->status);
```

Maybe something can be done if we afraid `pending_parts_completed` changed, but I've checked that it wouldn't happens

Reply:
`state->parts_in_progress` reached `0`, you unlock, another thread could increment `parts_in_progress` and now you call `MarkFinished`. Is `parts_in_progress > 0` considered "finished"?

Reply:
Ahhh, yes, it can. This is possible because the `Write` and `HandleUploadOutcome` can run concurrently, we'd better do like https://github.com/apache/arrow/pull/41876/files#r1631434845 to avoid mark finish before a task is really finished

Reply:
Um, you're right. @mapleFU's suggestion looks ok to me.

Note that the `pending_parts_completed` future can only be waited on in two situations:
* the user called blocking `Close` or `Flush` and the future is waited upon before returning from the API call;
* the user called non-blocking `CloseAsync`, which returns a cascaded future obtained by chaining `pending_parts_completed.Then` with a continuation.

For `Close` and `CloseAsync`, it's certainly not ok to call `Write` from another thread concurrently.
For `Flush`, it should be ok to call `Write` concurrently, but the `Flush` does not have to wait for the completion of the concurrent `Write` call.

Moreover, more generally, it doesn't seem sound to write to an output _stream_ (rather than random-access file) from several thread concurrently.

Reply:
> the user called blocking Close or Flush and the future is waited upon before returning from the API call;

Maybe there could be a sequence:
1. A last request finished, acquire lock, dec count, and set it to 0
2. New request sent, `pending_parts_completed` set to a new one
3. (1) call `pending_parts_completed.MarkFinished`, which may call on the new one

So the further blocking would wrong?